### PR TITLE
Allow adding bookmarks by url

### LINKED ISSUE #1628: Support adding bookmarks by url
Please remember to mention your version info (qutebrowser, Qt, PyQt,
OS/distribution) from the `qute:version` page or `qutebrowser --version`

---
(I am using Qutebrowser version 0.7.0)

# Summary
Is there any way of adding a bookmark by just entering the url?

# What I'm trying to do
I have a url (let's call it url1) that does some magic (probably with cookies) and then redirects to another webpage (let's call it url2). Doing `:bookmark-add` whilst on url2 does save url2 as a bookmark, but the magic that happens when url1 is clicked doesn't happen.

Just allowing bookmark-add to accept an optional url parameter would probs be the easiest way of doing this.

### PR DESCRIPTION
**Pull request by ismail-s**
Closes #1628.

### CODE REVIEW

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -1116,12 +1116,23 @@ def quickmark_load(self, name, tab=False, bg=False, window=False):
...
-    def bookmark_add(self):
-        """Save the current page as a bookmark."""
+    def bookmark_add(self, url=None, title=None):
+        """Save the current page as a bookmark, or a specific url."""
+        if url and not title:
+            raise cmdexc.CommandError('Title must be provided if url has '
+                                    'been provided')
         bookmark_manager = objreg.get('bookmark-manager')
-        url = self._current_url()
+        if isinstance(url, str):
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: What else would it be? I think you can simplify this whole block:

```python
if url is None:
    url = self._current_url()
else:
    try:
        url = urlutils.fuzzy_url(url)
    except urlutils.InvalidUrlError as e:
        raise cmdexc.CommandError(e)
```

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -1116,12 +1116,23 @@ def quickmark_load(self, name, tab=False, bg=False, window=False):
         self._open(url, tab, bg, window)
 
     @cmdutils.register(instance='command-dispatcher', scope='window')
-    def bookmark_add(self):
-        """Save the current page as a bookmark."""
+    def bookmark_add(self, url=None, title=None):
+        """Save the current page as a bookmark, or a specific url."""
+        if url and not title:
+            raise cmdexc.CommandError('Title must be provided if url has '
+                                    'been provided')
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Please indent this so the `'` match up

**Comment by The-Compiler** [comment]
Looks good to me! Can you please update the docstring and tests too?

**Comment by ismail-s** [comment]
That _should_ be everything done. Assuming the tests pass.

File: qutebrowser/browser/commands.py [hunk 3]
```diff
@@ -1116,12 +1116,32 @@ def quickmark_load(self, name, tab=False, bg=False, window=False):
...
+        """Save the current page as a bookmark, or a specific url.
+
+        If no url and title are provided, then save the current page as a
+        bookmark.
+        If a url and title have been provided, then save the given url as
+        a bookmark with the provided title.
+
+        Args:
+            url: url to save as a bookmark. If None, use url of current page.
+            title: title of the new bookmark."""
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: The `"""` should be on a separate line (as per pep8 or pep257, not sure which of those two coding standards :wink:)

Comment (ismail-s) [inline qutebrowser/browser/commands.py]: Yeah, ideally I would have run pylint on my machine but the tox command seemed to take ages (it's an old laptop I'm using).

**Comment by The-Compiler** [comment]
Sorry for the delay, and thanks!

**Pull request merged**